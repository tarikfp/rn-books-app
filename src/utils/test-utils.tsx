import { ThemeProvider } from '@shopify/restyle';
import { render, RenderOptions } from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import { theme } from '~theme/core';
import { ThemePreferenceContext } from '~theme/useThemePreference';

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <ThemePreferenceContext.Provider
      value={{
        isDarkTheme: false,
        toggleTheme: () => {},
      }}
    >
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </ThemePreferenceContext.Provider>
  );
};

const withAllTheProviders = (ui: React.ReactNode) => (
  <AllTheProviders children={ui} />
);

const customRender = (ui: React.ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AllTheProviders, ...options });

const customRenderer = (ui: React.ReactElement) =>
  renderer.create(withAllTheProviders(ui));

// re-export everything
export * from '@testing-library/react-native';
// override render method
export { customRender as render };
// override renderer method
export { customRenderer as createRenderer };
