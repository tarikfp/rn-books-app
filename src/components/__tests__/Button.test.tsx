import React from 'react';
import {
  BorderlessBounceButton,
  NativeBorderlessButton,
} from '~components/Button';
import { fireEvent, render } from '~utils/test-utils';

describe('NativeBorderlessButton Component', () => {
  it('matches snapshot with default props', () => {
    const { toJSON } = render(<NativeBorderlessButton label="Button Label" />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('matches snapshot when disabled', () => {
    const { toJSON } = render(
      <NativeBorderlessButton label="Button Label" enabled={false} />,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('triggers onPress mock when pressed', () => {
    const onPressMock = jest.fn();
    const { getByTestId } = render(
      <NativeBorderlessButton
        testID="button"
        label="Button Label"
        onPress={onPressMock}
      />,
    );

    fireEvent.press(getByTestId('button'));
    expect(onPressMock).toHaveBeenCalled();
  });

  it('triggers onPress mock when pressed and not disabled', () => {
    const onPressMock = jest.fn();
    const { getByTestId } = render(
      <NativeBorderlessButton
        testID="button"
        label="Button Label"
        onPress={onPressMock}
      />,
    );

    fireEvent.press(getByTestId('button'));
    expect(onPressMock).toHaveBeenCalled();
  });
});

describe('BorderlessBounceButton Component', () => {
  it('matches snapshot with default props', () => {
    const { toJSON } = render(
      <BorderlessBounceButton>
        <NativeBorderlessButton label="Button Label" />
      </BorderlessBounceButton>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('matches snapshot with custom bounceValue', () => {
    const { toJSON } = render(
      <BorderlessBounceButton bounceValue={0.8}>
        <NativeBorderlessButton label="Button Label" />
      </BorderlessBounceButton>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  const onPressMock = jest.fn();

  const { getByTestId } = render(
    <BorderlessBounceButton onPress={onPressMock}>
      <NativeBorderlessButton testID="button" label="Button Label" />
    </BorderlessBounceButton>,
  );

  fireEvent.press(getByTestId('button'));

  expect(onPressMock).toHaveBeenCalled();
});
