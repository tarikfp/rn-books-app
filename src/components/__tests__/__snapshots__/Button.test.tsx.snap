// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BorderlessBounceButton Component matches snapshot with custom bounceValue 1`] = `
<RNGestureHandlerButton
  activeOpacity={0.3}
  animatedStyle={
    {
      "value": {
        "transform": [
          {
            "scale": 1,
          },
        ],
      },
    }
  }
  borderless={true}
  collapsable={false}
  delayLongPress={600}
  handlerTag={9}
  handlerType="NativeViewGestureHandler"
  onActivated={[Function]}
  onActiveStateChange={[Function]}
  onEnded={[Function]}
  onGestureEvent={[Function]}
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  onHandlerStateChange={[Function]}
  style={
    {
      "opacity": 1,
      "transform": [
        {
          "scale": 1,
        },
      ],
    }
  }
>
  <RNGestureHandlerButton
    activeOpacity={0.3}
    animatedStyle={
      {
        "value": {
          "transform": [
            {
              "scale": 1,
            },
          ],
        },
      }
    }
    borderless={true}
    collapsable={false}
    delayLongPress={600}
    handlerTag={10}
    handlerType="NativeViewGestureHandler"
    onActivated={[Function]}
    onActiveStateChange={[Function]}
    onEnded={[Function]}
    onGestureEvent={[Function]}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onHandlerStateChange={[Function]}
    rippleRadius={0}
    style={
      {
        "opacity": 1,
        "transform": [
          {
            "scale": 1,
          },
        ],
      }
    }
  >
    <View
      style={
        [
          {
            "alignItems": "center",
            "backgroundColor": "#B4B6BB",
            "borderColor": "#E7E8E9",
            "borderRadius": 8,
            "borderWidth": 1,
            "justifyContent": "center",
            "paddingVertical": 12,
          },
        ]
      }
    >
      <Text
        numberOfLines={1}
        style={
          [
            {
              "color": "#FFFFFF",
              "fontFamily": "Inter-Medium",
              "fontSize": 14,
              "fontWeight": "800",
              "letterSpacing": -0.01,
              "lineHeight": 20,
            },
          ]
        }
      >
        Button Label
      </Text>
    </View>
  </RNGestureHandlerButton>
</RNGestureHandlerButton>
`;

exports[`BorderlessBounceButton Component matches snapshot with default props 1`] = `
<RNGestureHandlerButton
  activeOpacity={0.3}
  animatedStyle={
    {
      "value": {
        "transform": [
          {
            "scale": 1,
          },
        ],
      },
    }
  }
  borderless={true}
  collapsable={false}
  delayLongPress={600}
  handlerTag={7}
  handlerType="NativeViewGestureHandler"
  onActivated={[Function]}
  onActiveStateChange={[Function]}
  onEnded={[Function]}
  onGestureEvent={[Function]}
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  onHandlerStateChange={[Function]}
  style={
    {
      "opacity": 1,
      "transform": [
        {
          "scale": 1,
        },
      ],
    }
  }
>
  <RNGestureHandlerButton
    activeOpacity={0.3}
    animatedStyle={
      {
        "value": {
          "transform": [
            {
              "scale": 1,
            },
          ],
        },
      }
    }
    borderless={true}
    collapsable={false}
    delayLongPress={600}
    handlerTag={8}
    handlerType="NativeViewGestureHandler"
    onActivated={[Function]}
    onActiveStateChange={[Function]}
    onEnded={[Function]}
    onGestureEvent={[Function]}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onHandlerStateChange={[Function]}
    rippleRadius={0}
    style={
      {
        "opacity": 1,
        "transform": [
          {
            "scale": 1,
          },
        ],
      }
    }
  >
    <View
      style={
        [
          {
            "alignItems": "center",
            "backgroundColor": "#B4B6BB",
            "borderColor": "#E7E8E9",
            "borderRadius": 8,
            "borderWidth": 1,
            "justifyContent": "center",
            "paddingVertical": 12,
          },
        ]
      }
    >
      <Text
        numberOfLines={1}
        style={
          [
            {
              "color": "#FFFFFF",
              "fontFamily": "Inter-Medium",
              "fontSize": 14,
              "fontWeight": "800",
              "letterSpacing": -0.01,
              "lineHeight": 20,
            },
          ]
        }
      >
        Button Label
      </Text>
    </View>
  </RNGestureHandlerButton>
</RNGestureHandlerButton>
`;

exports[`NativeBorderlessButton Component matches snapshot when disabled 1`] = `
<RNGestureHandlerButton
  activeOpacity={0.3}
  animatedStyle={
    {
      "value": {
        "transform": [
          {
            "scale": 1,
          },
        ],
      },
    }
  }
  borderless={true}
  collapsable={false}
  delayLongPress={600}
  enabled={false}
  handlerTag={4}
  handlerType="NativeViewGestureHandler"
  onActivated={[Function]}
  onActiveStateChange={[Function]}
  onEnded={[Function]}
  onGestureEvent={[Function]}
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  onHandlerStateChange={[Function]}
  rippleRadius={0}
  style={
    {
      "opacity": 1,
      "transform": [
        {
          "scale": 1,
        },
      ],
    }
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#B4B6BB",
          "borderColor": "#E7E8E9",
          "borderRadius": 8,
          "borderWidth": 1,
          "justifyContent": "center",
          "paddingVertical": 12,
        },
      ]
    }
  >
    <Text
      numberOfLines={1}
      style={
        [
          {
            "color": "#FFFFFF",
            "fontFamily": "Inter-Medium",
            "fontSize": 14,
            "fontWeight": "800",
            "letterSpacing": -0.01,
            "lineHeight": 20,
          },
        ]
      }
    >
      Button Label
    </Text>
  </View>
</RNGestureHandlerButton>
`;

exports[`NativeBorderlessButton Component matches snapshot with default props 1`] = `
<RNGestureHandlerButton
  activeOpacity={0.3}
  animatedStyle={
    {
      "value": {
        "transform": [
          {
            "scale": 1,
          },
        ],
      },
    }
  }
  borderless={true}
  collapsable={false}
  delayLongPress={600}
  handlerTag={3}
  handlerType="NativeViewGestureHandler"
  onActivated={[Function]}
  onActiveStateChange={[Function]}
  onEnded={[Function]}
  onGestureEvent={[Function]}
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  onHandlerStateChange={[Function]}
  rippleRadius={0}
  style={
    {
      "opacity": 1,
      "transform": [
        {
          "scale": 1,
        },
      ],
    }
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#B4B6BB",
          "borderColor": "#E7E8E9",
          "borderRadius": 8,
          "borderWidth": 1,
          "justifyContent": "center",
          "paddingVertical": 12,
        },
      ]
    }
  >
    <Text
      numberOfLines={1}
      style={
        [
          {
            "color": "#FFFFFF",
            "fontFamily": "Inter-Medium",
            "fontSize": 14,
            "fontWeight": "800",
            "letterSpacing": -0.01,
            "lineHeight": 20,
          },
        ]
      }
    >
      Button Label
    </Text>
  </View>
</RNGestureHandlerButton>
`;
